data-prepper-plugins/rds-source/src/main/java/org/opensearch/dataprepper/plugins/source/rds/schema/PostgresSchemaManager.java

```diff
@@ -0,0 +1,109 @@
+        StringBuilder createPublicationStatementBuilder = new StringBuilder("CREATE PUBLICATION ")
+                .append(publicationName)
+                .append(" FOR TABLE ");
+        for (int i = 0; i < tableNames.size(); i++) {
+            createPublicationStatementBuilder.append(tableNames.get(i));
+            if (i < tableNames.size() - 1) {
+                createPublicationStatementBuilder.append(", ");
+            }
+        }
+        createPublicationStatementBuilder.append(";");
+        final String createPublicationStatement = createPublicationStatementBuilder.toString();
+
+        try (Connection conn = connectionManager.getConnection()) {
+            try {
+                PreparedStatement statement = conn.prepareStatement(createPublicationStatement);
+                statement.executeUpdate();
+            } catch (Exception e) {
+                LOG.info("Failed to create publication: {}", e.getMessage());
+            }
+
+            PGConnection pgConnection = conn.unwrap(PGConnection.class);
+
+            // Create replication slot
+            PGReplicationConnection replicationConnection = pgConnection.getReplicationAPI();
+            try {
+                replicationConnection.createReplicationSlot()
+                        .logical()
+                        .withSlotName(slotName)
+                        .withOutputPlugin("pgoutput")
+                        .make();
+                LOG.info("Replication slot {} created successfully. ", slotName);
+            } catch (Exception e) {
+                LOG.info("Failed to create replication slot {}: {}", slotName, e.getMessage());
+            }
+        } catch (Exception e) {
+            LOG.error("Exception when creating replication slot. ", e);
+        }
+    }
+
+    @Override
+    public List<String> getPrimaryKeys(final String fullTableName) {
+        final String[] splits = fullTableName.split("\\.");
+        final String database = splits[0];
+        final String schema = splits[1];
+        final String table = splits[2];
+        int retry = 0;
+        while (retry <= NUM_OF_RETRIES) {
+            final List<String> primaryKeys = new ArrayList<>();
+            try (final Connection connection = connectionManager.getConnection()) {
+                try (final ResultSet rs = connection.getMetaData().getPrimaryKeys(database, schema, table)) {
+                    while (rs.next()) {
+                        primaryKeys.add(rs.getString(COLUMN_NAME));
+                    }
+                    return primaryKeys;
+                }
+            } catch (Exception e) {
+                LOG.error("Failed to get primary keys for table {}, retrying", table, e);
+            }
+            applyBackoff();
+            retry++;
+        }
+        LOG.warn("Failed to get primary keys for table {}", table);
+        return List.of();
```

Comment:
Exception handling when PrimaryKey is not present vs Internal DB Exception/Errors.

Reply:
Added.